internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -507,6 +510,113 @@ func (a *Adapter) EnsureOverrideSnapshotValid() (controller.OperationResult, err
+func (a *Adapter) EnsureGroupSnapshotExist() (controller.OperationResult, error) {
+	if gitops.IsSnapshotCreatedByPACPushEvent(a.snapshot) {
+		a.logger.Info("The snapshot is not created by PAC pull request, no need to create group snapshot")
+		return controller.ContinueProcessing()
+	}
+
+	if !metadata.HasLabelWithValue(a.snapshot, gitops.SnapshotTypeLabel, gitops.SnapshotComponentType) {
+		a.logger.Info("The snapshot is not a component snapshot, no need to create group snapshot for it")
+		return controller.ContinueProcessing()
+	}
+
+	if gitops.HasPRGroupProcessed(a.snapshot) {
+		a.logger.Info("The PR group info has been processed for this component snapshot, no need to process it again")
+		return controller.ContinueProcessing()
+	}
+
+	prGroupHash := gitops.GetPRGroupHashFromSnapshot(a.snapshot)
+	if prGroupHash == "" {
+		a.logger.Error(fmt.Errorf("NotFound"), fmt.Sprintf("Failed to get PR group hash from snapshot %s/%s", a.snapshot.Namespace, a.snapshot.Name))
+		err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, fmt.Sprintf("Failed to get PR group hash from snapshot %s/%s", a.snapshot.Namespace, a.snapshot.Name), a.client)
+		if err != nil {
+			return controller.RequeueWithError(err)
+		}
+		return controller.ContinueProcessing()
+	}
+
+	prGroup := gitops.GetPRGroupFromSnapshot(a.snapshot)
+	if prGroup == "" {
+		a.logger.Error(fmt.Errorf("NotFound"), fmt.Sprintf("Failed to get PR group from snapshot %s/%s", a.snapshot.Namespace, a.snapshot.Name))
+		err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, fmt.Sprintf("Failed to get PR group from snapshot %s/%s", a.snapshot.Namespace, a.snapshot.Name), a.client)
+		if err != nil {
+			return controller.RequeueWithError(err)
+		}
+		return controller.ContinueProcessing()
+	}
+
+	pipelineRuns, err := a.loader.GetPipelineRunsWithPRGroupHash(a.context, a.client, a.snapshot, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, fmt.Sprintf("Failed to get build pipelineRuns for given pr group hash %s", prGroupHash))
+		return controller.RequeueWithError(err)
+	}
+
+	for _, pipelineRun := range *pipelineRuns {
+		pipelineRun := pipelineRun
+
+		// check if build PLR finishes
+		if !h.HasPipelineRunFinished(&pipelineRun) {
+			a.logger.Info(fmt.Sprintf("The build pipelineRun %s/%s with pr group %s is still running, won't create group snapshot", pipelineRun.Namespace, pipelineRun.Name, prGroup))
+			err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, fmt.Sprintf("The build pipelineRun %s/%s with pr group %s is still running, won't create group snapshot", pipelineRun.Namespace, pipelineRun.Name, prGroup), a.client)
+			if err != nil {
+				return controller.RequeueWithError(err)
+			}
+			return controller.ContinueProcessing()
+		}
+
+		// check if build PLR succeeds
+		if !h.HasPipelineRunSucceeded(&pipelineRun) {
```

Comment:
Will this block creating a PR group Snapshot even if the build pipeline for that component is rerun and then succeeds?

Reply:
You are right. Thanks! A function `isLatestBuildPipelineRunInComponent()` is added at the beginning of this PLRs loop.